### PR TITLE
Add red53_agent

### PR DESCRIPTION
**Pull request by mishan**
- This service is to be run as a singleton. It will periodically check a Redis ZSET and modify Route53 DNS records accordingly

@TysonAndre

### CODE REVIEW

File: red53_agent.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+					},
+				},
+			},
+			Comment: aws.String("Service Discovery Created Record"),
+		},
+		HostedZoneId: aws.String(configuration.HostedZoneId),
+	}
+	_, err = r53.ChangeResourceRecordSets(params)
+	logErrorNoFatal(err)
+	log.Debug("Record " + serviceName + "." + DNSName + " updated.")
```
Comment (TysonAndre) [inline red53_agent.go]: log number of ip addresses?

File: red53_agent.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+const checkInterval = 5 * time.Minute // how often to check Redis
+const defaultTTL = 60                 // seconds
+const fetchLast = 300                 // seconds
+
+var DNSName = "servicediscovery.local"
+
+func logErrorAndFail(err error) {
+	if err != nil {
+		log.Fatal(err)
```
Comment (TysonAndre) [inline red53_agent.go:31]: Add comment that this causes process to exit() in logrus

File: red53_agent.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+		}
+		if sum > 8 {
+			logErrorAndFail(err)
+		}
+		time.Sleep(time.Duration(sum) * time.Second)
+		sum += 2
+	}
+
+	// check regularly, specified by checkInterval
+	ticker := time.NewTicker(checkInterval)
```
Comment (TysonAndre) [inline red53_agent.go:198]: May be possible to use pubsub, if aws supports that.

Could check ZAdd on the publishing side to see if anything was added (it returns 1 if it didn't already exist)

> ZADD Return:    The number of elements added to the sorted sets, not including elements already existing for which the score was updated.

Then call publish() on the ZADD caller

Call PSubscribe with a channel, here

Then listen with `select` on both the ticker and the redis pubsub channel.

File: red53_agent.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
+package main
+
+// Originally adapted from https://github.com/awslabs/service-discovery-ecs-dns
+// This service is meant to run as a singleton instance in concert with redecs_agent
+// running on AWS ECS instances. This service will check Redis regularly for active
+// service updates and rewrite the A records for services to only contain the most
```
Comment (TysonAndre) [inline red53_agent.go]: nit: to only contain recently active IP address

File: red53_agent.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+		servicePings, err = redisClient.ZRange("redecs:service_pings", epoch, now).Result()
+
+		if err == nil {
+			break
+		}
+		if sum > 8 {
+			// Bail out if this is failing.
+			logErrorAndFail(err)
+		}
+		time.Sleep(time.Duration(sum) * time.Second)
```
Comment (TysonAndre) [inline red53_agent.go:138]: Confusing cast: time.Duration(sum) isn't a meaningful duration. sum \* time.Second should already be a duration?

time.Duration(sum) would technically be 5 nanoseconds?

**Review by TysonAndre — COMMENTED** [review]
Seems fine overall. Also, what happens the invoker of this script detects that it failed/ended? Does it restart the script or log to anywhere visible?

File: red53_agent.go (hunk 3 above)
Comment (mishan) [reply red53_agent.go:198]: +1 Yeah. can definitely use pub sub, was going to do that after initial POC

Comment (mishan) [reply red53_agent.go:198]: Guess can't hurt to add here now in advance. :)

Comment (mishan) [reply red53_agent.go:198]: This part is much easier to swap out than redecs_agent which I just deployed, so I'll add a TODO here for this enhancement.

**Comment by mishan** [comment]
I plan to run it in a Docker container, so the container would start up again in ECS. Re: logging, we don't have the centralized logging infrastructure yet, but that will come soon. I could put logs in the EFS mount for the meantime

File: red53_agent.go (hunk 5 above)
Comment (mishan) [reply red53_agent.go:138]: https://golang.org/pkg/time/#Duration

seconds := 10
fmt.Print(time.Duration(seconds)*time.Second) // prints 10s

I agree it's weird.

**Comment by TysonAndre** [comment]
:shipit:

**Pull request merged**